x-pack/plugins/task_manager/task_store.ts

```diff
@@ -0,0 +1,370 @@
+ */
+
+import { ConcreteTaskInstance, ElasticJs, TaskInstance, TaskStatus } from './task';
+
+const DOC_TYPE = '_doc';
+
+export interface StoreOpts {
+  callCluster: ElasticJs;
+  index: string;
+  maxAttempts: number;
+  supportedTypes: string[];
+}
+
+export interface FetchOpts {
+  searchAfter?: any[];
+  sort?: object[];
+  query?: object;
+}
+
+export interface FetchResult {
+  searchAfter: any[];
+  docs: ConcreteTaskInstance[];
+}
+
+export interface RemoveResult {
+  index: string;
+  id: string;
+  version: string;
+  result: string;
+}
+
+// Internal, the raw document, as stored in the Kibana index.
+export interface RawTaskDoc {
+  _id: string;
+  _index: string;
+  _type: string;
+  _version: number;
+  _source: {
+    type: string;
+    task: {
+      taskType: string;
+      runAt: Date;
+      interval?: string;
+      attempts: number;
+      status: TaskStatus;
+      params: string;
+      state: string;
+      user?: string;
+      scope?: string | string[];
+    };
+  };
+}
+
+/**
+ * Wraps an elasticsearch connection and provides a task manager-specific
+ * interface into the index.
+ */
+export class TaskStore {
+  public readonly maxAttempts: number;
+  private callCluster: ElasticJs;
+  private index: string;
+  private supportedTypes: string[];
+  private wasInitialized = false;
+
+  /**
+   * Constructs a new TaskStore.
+   * @param {StoreOpts} opts
+   * @prop {CallCluster} callCluster - The elastic search connection
+   * @prop {string} index - The name of the task manager index
+   * @prop {number} maxAttempts - The maximum number of attempts before a task will be abandoned
+   * @prop {string[]} supportedTypes - The task types supported by this store
+   */
+  constructor(opts: StoreOpts) {
+    this.callCluster = opts.callCluster;
+    this.index = opts.index;
+    this.maxAttempts = opts.maxAttempts;
+    this.supportedTypes = opts.supportedTypes;
+  }
+
+  public addSupportedTypes(types: string[]) {
+    if (!this.wasInitialized) {
+      this.supportedTypes = this.supportedTypes.concat(types);
+    } else {
+      throw Error('Cannot add task types after initializattion');
+    }
+  }
+
+  /**
+   * Initializes the store, ensuring the task manager index is created and up to date.
+   */
+  public async init() {
+    if (this.wasInitialized) {
+      return;
+    }
+
+    const properties = {
+      type: { type: 'keyword' },
+      task: {
+        properties: {
+          taskType: { type: 'keyword' },
+          runAt: { type: 'date' },
+          interval: { type: 'text' },
+          attempts: { type: 'integer' },
+          status: { type: 'keyword' },
+          params: { type: 'text' },
+          state: { type: 'text' },
+          user: { type: 'keyword' },
+          scope: { type: 'keyword' },
+        },
+      },
+    };
+
+    try {
+      const templateResult = await this.callCluster('indices.putTemplate', {
+        name: this.index,
+        body: {
+          index_patterns: [this.index],
```

Comment:
It worries me to see us doing this here rather than reusing some core Kibana library to do this (I'm not saying that one does exist, but one _should_ exist). This template is so generic that we are hopefully able to just set it via a very wide index pattern, like `.kibana_tasks_*` (assuming that matches the index's default name), rather than trying to create a template per index name.

But also we should not be setting this unless we need to set it. If it already exists, then we're unnecessary causing a cluster state change on the ES side every time that Kibana starts up.

As an aside, the ES Template API supports `version` values and we should be using it based on the version of Kibana so that we do not replace a _newer_ version of Kibana's template with our own. And, given a newer template, we should consider not marking the task manager as initialized.